File: day_2/part1.js
const fs = require("fs");

const parseCommand = (command) => {
  const distance = parseInt(command.split(" ")[1]) || 0;
  if (command.includes("forward")) {
    return {
      depth: 0,
      horizontal: distance,
    };
  }
  if (command.includes("down")) {
    return {
      depth: distance,
      horizontal: 0,
    };
  }
  return {
    depth: -distance,
    horizontal: 0,
  };
};
const instructions = fs
  .readFileSync("input.txt")
  .toString()
  .split("\n")
  .map(parseCommand);

const endPosition = instructions.reduce(
  ({ depth, horizontal }, currCommand) => ({
    depth: depth + currCommand.depth,
    horizontal: horizontal + currCommand.horizontal,
  }),
  {
    depth: 0,
    horizontal: 0,
  }
);
console.log(endPosition);
console.log(endPosition.depth * endPosition.horizontal);

File: day_2/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Vector struct {
	depth      int
	horizontal int
}

func (v1 Vector) add(v2 Vector) Vector {
	return Vector{
		depth:      v1.depth + v2.depth,
		horizontal: v1.horizontal + v2.horizontal,
	}
}

func parseVector(command string) Vector {
	parts := strings.Split(command, " ")
	commandType := parts[0]
	distance, _ := strconv.Atoi(parts[1])
	if commandType == "forward" {
		return Vector{depth: 0, horizontal: distance}
	}
	if commandType == "down" {
		return Vector{depth: distance, horizontal: 0}
	}
	return Vector{depth: -distance, horizontal: 0}
}

func main() {
	f, err := os.Open("input.txt")
	if err != nil {
		fmt.Printf("error opening file: %v\n", err)
		os.Exit(1)
	}
	scanner := bufio.NewScanner(f)
	currentPosition := Vector{depth: 0, horizontal: 0}
	for scanner.Scan() {
		vector := parseVector(scanner.Text())
		currentPosition = currentPosition.add(vector)
	}
	fmt.Println(currentPosition.horizontal * currentPosition.depth)
}
